### PR TITLE
Microchip: MEC172x: QMSPI driver

### PR DESCRIPTION
**Pull request by scottwcpg**
Create a separate QMSPI driver for MEC172x as the HW has
been enhanced with extra timing registers and local DMA channels.

Signed-off-by: Scott Worley <scott.worley@microchip.com>

### CODE REVIEW

File: drivers/spi/Kconfig.xec_qmspi_v2 [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+# Copyright (c) 2021 Microchip Technology Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+config SPI_XEC_QMSPI_V2
+	bool "Microchip XEC QMSPI driver"
+	default y
+	depends on SOC_SERIES_MEC172X
+	select DMA if SPI_ASYNC
+	help
+	  Enable support for the Microchip XEC QMSPI V2 driver.
```
Comment (jackrosenthal) [inline drivers/spi/Kconfig.xec_qmspi_v2:12]: expand acronyms

File: drivers/spi/spi_xec_qmspi_v2.c [hunk 2]
```diff
@@ -0,0 +1,751 @@
...
+ * Allocate(build) one or more descriptors.
+ * QMSPI contains 16 32-bit descriptor registers used as a linked
+ * list of operations. Using only 32-bits there are limitations.
+ * Each descriptor is limited to 0x7FFF units where unit size can
+ * be 1, 4, or 16 bytes. A descriptor can perform transmit or receive
+ * but not both simultaneously. Order of descriptor processing is specified
+ * by the first descriptor field of the control register, the next descriptor
+ * fields in each descriptor, and the descriptors last flag.
+ */
+static int qmspi_descr_alloc(struct qmspi_regs *regs, const struct spi_buf *txb,
```
Comment (albertofloyd) [inline drivers/spi/spi_xec_qmspi_v2.c:361]: @scottwcpg  could you document the parameters of this internal function?  I'm assuming didx is the descriptor index 
if that's the case, maybe rework the parameter list and pass didx as reference.
Returning an error and also valid value may lead to error for future maintainers. 
Seems didx = 0 is not equivalent to success in all scenarios.

File: drivers/spi/Kconfig.xec_qmspi_v2 [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# Microchip XEC QMSPI V2
+
+# Copyright (c) 2021 Microchip Technology Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+config SPI_XEC_QMSPI_V2
+	bool "Microchip XEC QMSPI driver"
+	default y
+	depends on SOC_SERIES_MEC172X
+	select DMA if SPI_ASYNC
```
Comment (jackrosenthal) [inline drivers/spi/Kconfig.xec_qmspi_v2:10]: select on visible symbols should be avoided

https://docs.zephyrproject.org/latest/guides/build/kconfig/tips.html#select-pitfalls

File: drivers/spi/Kconfig.xec_qmspi_v2 [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# Microchip XEC QMSPI V2
+
+# Copyright (c) 2021 Microchip Technology Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+config SPI_XEC_QMSPI_V2
```
Comment (jackrosenthal) [inline drivers/spi/Kconfig.xec_qmspi_v2:6]: what is "v2" for?

the help does not explain it

File: drivers/spi/spi_xec_qmspi_v2.c [hunk 5]
```diff
@@ -0,0 +1,751 @@
...
+ */
+static int qmspi_tx_dummy_clocks(struct qmspi_regs *regs, uint32_t nclocks)
+{
+	uint32_t descr, ifm, qstatus;
+
+	ifm = regs->CTRL & MCHP_QMSPI_C_IFM_MASK;
+	descr = ifm | MCHP_QMSPI_C_TX_DIS | MCHP_QMSPI_C_XFR_UNITS_BITS
+		| MCHP_QMSPI_C_DESCR_LAST | MCHP_QMSPI_C_DESCR0;
+
+	if (ifm & 0x01) {
```
Comment (albertofloyd) [inline drivers/spi/spi_xec_qmspi_v2.c:290]: Magic number

File: boards/arm/mec172xevb_assy6906/pinmux.c [hunk 6]
```diff
@@ -80,6 +80,61 @@ static void brd_cfg_uart(struct pinmux_ports_t *pp)
...
+static void brd_cfg_qmspi(struct pinmux_ports_t *pp)
+{
+	pinmux_pin_set(pp->porta, MCHP_GPIO_024, MCHP_GPIO_CTRL_MUX_F1 |
+						 MCHP_GPIO_CTRL_BUFT_OPENDRAIN);
+	pinmux_pin_set(pp->porta, MCHP_GPIO_023, MCHP_GPIO_CTRL_MUX_F1);
+	pinmux_pin_set(pp->portf, MCHP_GPIO_245, MCHP_GPIO_CTRL_MUX_F1);
+	pinmux_pin_set(pp->portf, MCHP_GPIO_243, MCHP_GPIO_CTRL_MUX_F1);
+}
+#else
+BUILD_ASSERT(0, "QMSPI DT port_del illegal value!");
```
Comment (jackrosenthal) [inline boards/arm/mec172xevb_assy6906/pinmux.c:129]: #error

File: boards/arm/mec172xevb_assy6906/pinmux.c [hunk 7]
```diff
@@ -80,6 +80,61 @@ static void brd_cfg_uart(struct pinmux_ports_t *pp)
 #endif
 }
 
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(spi0), okay)
+
+#if DT_PROP(DT_INST(0, microchip_xec_qmspi), port_sel) == 0
+/* shared QMSPI port: Two chip selects, single, dual, or quad */
+static void brd_cfg_qmspi(struct pinmux_ports_t *pp)
+{
+#if DT_PROP(DT_INST(0, microchip_xec_qmspi), chip_select) == 0
```
Comment (jackrosenthal) [inline boards/arm/mec172xevb_assy6906/pinmux.c:89]: can the gpio port/pin be retreived from device tree so we don't need this ifdef?

**Comment by scottwcpg** [comment]
Soon to transition to new maintainer

**Pull request closed without merging**